tools/ci/mkchlog-snippet.py

```diff
@@ -0,0 +1,210 @@
+                + f" via {self.as_link()} (by {self.author.as_link()})"
+            )
+        else:
+            item += f"  {self.as_link()} (by {self.author.as_link()})"
+        item = textwrap.fill(
+            item,
+            width=79,
+            initial_indent="- ",
+            subsequent_indent="  ",
+            break_long_words=False,
+            break_on_hyphens=False,
+        )
+        return f"### {self.category}\n\n{item}\n"
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-d", "--outdir", default="changelog.d", type=Path)
+    parser.add_argument("repo_owner")
+    parser.add_argument("repo_name")
+    parser.add_argument("prnum", type=int)
+    args = parser.parse_args()
+    try:
+        token = os.environ["GITHUB_TOKEN"]
+    except KeyError:
+        sys.exit("GITHUB_TOKEN not set")
+    pr = get_pr_info(
+        repo_owner=args.repo_owner,
+        repo_name=args.repo_name,
+        prnum=args.prnum,
+        token=token,
+    )
+    args.outdir.mkdir(parents=True, exist_ok=True)
+    outfile = args.outdir / f"pr-{args.prnum}.md"
+    outfile.write_text(pr.as_snippet())
+    print("Changelog snippet saved to", outfile)
+
+
+def get_pr_info(repo_owner: str, repo_name: str, prnum: int, token: str) -> PullRequest:
+    q = (
+        "query(\n"
+        "  $repo_owner: String!,\n"
+        "  $repo_name: String!,\n"
+        "  $prnum: Int!,\n"
+        "  $closing_cursor: String,\n"
+        "  $label_cursor: String\n"
+        ") {\n"
+        "  repository(owner: $repo_owner, name: $repo_name) {\n"
+        "    pullRequest(number: $prnum) {\n"
+        "      title\n"
+        "      number\n"
+        "      url\n"
+        "      author {\n"
+        "        login\n"
+        "        url\n"
+        "      }\n"
+        "      closingIssuesReferences(\n"
+        "        first: 50,\n"
+        "        orderBy: {field: CREATED_AT, direction:ASC},\n"
+        "        after: $closing_cursor\n"
+        "      ) {\n"
+        "        nodes {\n"
+        "          number\n"
+        "          url\n"
+        "        }\n"
+        "        pageInfo {\n"
+        "          endCursor\n"
+        "          hasNextPage\n"
+        "        }\n"
+        "      }\n"
+        "      labels(first: 50, after: $label_cursor) {\n"
+        "        nodes {\n"
+        "          name\n"
+        "        }\n"
+        "        pageInfo {\n"
+        "          endCursor\n"
+        "          hasNextPage\n"
+        "        }\n"
+        "      }\n"
+        "    }\n"
+        "  }\n"
+        "}\n"
+    )
+    variables = {
+        "repo_owner": repo_owner,
+        "repo_name": repo_name,
+        "prnum": prnum,
+        "closing_cursor": None,
+        "label_cursor": None,
+    }
+    closed_issues: list[str] = []
+    labels: set[str] = set()
+    with requests.Session() as s:
+        s.headers["Authorization"] = f"bearer {token}"
+        while True:
```

Comment:
This loop necessity/logic is not clear to me: from reading the code - it loops if there are next pages for issues and/or labels, while overwriting `data` (used to get metadata for the PR itself) on next iteration.  
Does every "page" repeat all the same information about PR (so overloading of `data` with title/number/url would be ok) but with some extra metadata for issues/labels?  if so -- ok, if not -- might need fixup of logic.

In either case (either logic needs to be fixed or not), @jwodder could you please add some comment above the `while True` loop describing why it is needed?

Reply:
Thanks for the explanation and the comment. As long as it works -- fine with me as-is given that we should not get over 50 in any of those "categories" ;)